### PR TITLE
Replace prob based with threshold based load balancing 

### PR DESCRIPTION
**Pull request by ByronHsu**
## Motivation

Prob based LB can disturb cache aware when the load is actually balanced. Switching to threshold based LB can help improve the cache hit rate and also makes the behavior more deterministic.

## Modifications

1. `bench_serving`: for generated shared prefix dataset, Remove argument-based data caching and use auto caching with the key as the dataset params, so users don't have to manually configure the arguments
2. Shortest queue load balancing is triggered when the load is imbalanced. The imbalanced is defined as `(max_load - min_load) > abs_threshold AND max_load > min_load * rel_threshold`. By default it only uses `abs_threshold`, and `rel_threshold` was set to a very small value.

## Benchmark Results

TLDR: the perf is on par with v1 on perfectly divided and low hit rate cases. In one long sys prompt case, it can outperform original RR DP, while it is slower than v1 because one worker will have "abs_threshold" more requests.

### Generated Shared Prefix Dataset
```bash
python bench_serving.py --host 127.0.0.1 --port 30000 --dataset-name generated-shared-prefix \
    --generated-input-path ~/.cache/gen.json --generated-input-save-path ~/.cache/gen.json
```

| Method | Throughput | Cache Rate |
|--------|------------|------------|
| Original RR DP | 82,665 | 20% |
| Cache Aware v1 | 158,596.72 | 75% |
| Perfect | 160,288 | 75% |
| **Cache Aware v1.1** | 158554 | 75% |

### SharedGPT Dataset
```bash
python bench_serving.py --host 127.0.0.1 --port 30000
```

| Method | Throughput | Cache Rate |
|--------|------------|------------|
| Original RR DP | 17,164 | 2% |
| Cache Aware v1 | 17,775 | 2% |
| **Cache Aware v1** | 17,779 | 2% |

### Multi Turn Dataset
```bash
python long_prompt_multi_turn.py --port 30000 --tokenizer "/shared/public/elr-models/meta-llama/Meta-Llama-3.1-8B-Instruct/07eb05b21d191a58c577b4a45982fe0c049d0693/" | tee client.log
```

| Method | Latency | Cache Rate |
|--------|---------|------------|
| Original RR DP | 34 | 35% |
| Cache Aware v1 | 19 | 88% |
| Perfect | 19 | 88% |
| **Cache Aware v1.1** | 19 | 88% |

### Generated Shared Prefix Dataset but only has one system prompt
```bash
python bench_serving.py --host 127.0.0.1 --port 30000 --dataset-name generated-shared-prefix --gen-num-groups 8 --gen-num-groups 1 --gen-prompts-per-group 1024
```

| Version | Throughput | Cache Rate |
|---------|------------|------------|
| Original RR DP | 154535.56 | |
| Cache aware v1 | 36510.71 | |
| Cache aware v1 - routing prob 0.5 | 190026.64 | |
| **Cache aware v1.1** | 174807 | |

## Checklist

- [x] Format your code according to the [Contributor Guide](https://github.com/sgl-project/sglang/blob/main/docs/references/contributor_guide.md).
- [x] Add unit tests as outlined in the [Contributor Guide](https://github.com/sgl-project/sglang/blob/main/docs/references/contributor_guide.md).
- [x] Update documentation as needed, including docstrings or example tutorials.

### CODE REVIEW

**Comment by ByronHsu** [comment]
TODO:
1. Replace the value in running_queue with text.chars().count() => so it can reflect the actual #char workload
2. Further tune the data value based on real world workload

**Review by Ying1123 — APPROVED** [review]

**Comment by ByronHsu** [comment]
https://github.com/sgl-project/sglang/issues/1732
